qiskit/circuit/library/n_local/qaoa_ansatz.py

```diff
@@ -0,0 +1,257 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2018, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A generalized QAOA quantum circuit with a support of custom initial states and mixers."""
+# pylint: disable=R0401
+from typing import Optional, cast, Set
+
+import numpy as np
+
+from qiskit.circuit.parameter import Parameter
+from qiskit.circuit.parametervector import ParameterVector
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from ..blueprintcircuit import BlueprintCircuit
+
+
+class QAOAAnsatz(BlueprintCircuit):
+    """A generalized QAOA quantum circuit with a support of custom initial states and mixers.
+
+    References:
+
+        [1]: Farhi et al., A Quantum Approximate Optimization Algorithm.
+            `arXiv:1411.4028 <https://arxiv.org/pdf/1411.4028>`_
+    """
+    def __init__(self,
+                 cost_operator=None,
+                 reps: int = 1,
+                 initial_state: Optional[QuantumCircuit] = None,
+                 mixer_operator=None,
+                 name: str = "qaoa"):
+        r"""
+        Args:
+            cost_operator (OperatorBase, optional): The operator representing the cost of
+                the optimization problem, denoted as :math:`U(C, \gamma)` in the original paper.
+                Must be set either in the constructor or via property setter.
+            reps (int): The integer parameter p, which determines the depth of the circuit,
+                as specified in the original paper, default is 1.
+            initial_state (QuantumCircuit, optional): An optional initial state to use.
+                If `None` is passed then a set of Hadamard gates is applied as an initial state
+                to all qubits.
+            mixer_operator (OperatorBase or QuantumCircuit, optional): An optional custom mixer
+                to use instead of the global X-rotations, denoted as :math:`U(B, \beta)`
+                in the original paper. Can be an operator or an optionally parameterized quantum
+                circuit.
+            name (str): A name of the circuit, default 'qaoa'
+        """
+        super().__init__(name=name)
+        self._cost_operator = cost_operator
+        self._reps = reps
+        self._initial_state = initial_state
+        self._mixer = mixer_operator
+
+        # set this circuit as a not-built circuit
+        self._num_qubits = cost_operator.num_qubits if cost_operator else None
+        self._num_parameters = 0
+        self._bounds = None
```

Comment:
Now they are in use, honestly, I don't know how correct this implementation is. VQE and BlueprintCircuit in conjunction is a total mess. Since they were not used, the results of some tests are different now.

Reply:
If you'd ask me I'm not sure if parameter bounds should be attached to a circuit, the optimization might come out differently e.g. if you choose `(-pi,pi)` or `(0, 2pi)` or maybe you want to include an epsilon more or so... But we could talk about the whole Blueprint + VQE workflow (after this PR)!

Reply:
For a more general bounds discussion also see #5713